File: app/src/main/cpp/utils/files.cpp
/*
 * Copyright 2023 Michele Borassi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <experimental/filesystem>

#include "files.h"

void CreateFileIfNotExists(std::string filepath) {
  if (std::fstream(filepath, std::ios::in).is_open()) {
    return;
  }
  ResetFile(filepath);
}


void ResetFile(std::string filepath) {
  auto path = std::experimental::filesystem::path(filepath);
  std::experimental::filesystem::create_directories(path.remove_filename());
  std::ofstream(filepath, std::ios::out).close();
//  file_ = std::fstream(filepath_, std::ios::binary | std::ios::out | std::ios::in);
//  file_.seekg(0, std::ios::end);
//  assert(file_.is_open());
}

FileOffset FileLength(std::fstream& file) {
  assert(file.is_open());
  auto position = file.tellg();
  file.seekg(0, std::ios_base::end);
  auto result = file.tellg();
  file.seekg(position, std::ios_base::beg);
  return result;
}

File: lib/files.dart
/*
 * Copyright (c) 2023-2025 Michele Borassi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:othello_sensei/state.dart';
import 'package:path/path.dart';

const String assetVersion = "14";

String localAssetPath() {
  return join(GlobalState.localPath, "assets");
}

String localSavedGamesFoldersPath() {
  return join(GlobalState.localPath, "saved_games_folders.txt");
}

String localTempPath() {
  return join(GlobalState.localPath, "tmp");
}

String versionFilePath() {
  return join(localAssetPath(), "version.txt");
}

Future<void> copyAssetsToLocalPath() async {
  final assetManifest = await AssetManifest.loadFromAssetBundle(rootBundle);
  final assets = assetManifest.listAssets();
  var directory = Directory(localAssetPath());
  if (directory.existsSync()) {
    directory.deleteSync(recursive: true);
  }
  for (var path in assets) {
    // Avoid copying .DS_Store in Mac as it causes a failure.
    if (basename(path).startsWith('.')) {
      continue;
    }
    var destinationPath = join(GlobalState.localPath, path);
    var destinationDirName = dirname(destinationPath);
    Directory(destinationDirName).createSync(recursive: true);
    final encryptedByteData = await rootBundle.load(path);
    File(destinationPath).writeAsBytesSync(encryptedByteData.buffer.asUint8List());
  }
  File(versionFilePath()).writeAsString(assetVersion);
}

Future<void> maybeCopyAssetsToLocalPath() async {
  String storedAssetVersion = '';
  try {
    storedAssetVersion = File(versionFilePath()).readAsStringSync();
  } on PathNotFoundException {}
  if (assetVersion != storedAssetVersion) {
    await copyAssetsToLocalPath();
  }
}